src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/LinkedInversionBreakpointsInference.java

```diff
@@ -0,0 +1,935 @@
+
+        return overlappingPairs.stream()
+                .map(LinkedInversionBreakpointsInference::factory)
+                .flatMap(overlappingPairHandler ->
+                        overlappingPairHandler.toVariants(reference, assemblyID2Reads, pathToFastqsDir).stream())
+                .collect(Collectors.toList());
+    }
+
+    //==================================================================================================================
+
+    private abstract static class OverlappingPairHandler implements Serializable {
+        private static final long serialVersionUID = 1L;
+
+        protected final OverlappingPair overlappingPair;
+
+        OverlappingPairHandler(final OverlappingPair overlappingPair) {
+            this.overlappingPair = overlappingPair;
+        }
+
+        /**
+         * Inference logic to be implemented case by case.
+         * @param reference         reference for extracting relevant bases
+         * @param assemblyID2Reads  map from assembly ID to short reads used for its assembly process
+         * @param pathToFastqsDir   a development artifact, because we want to save alignment results to file and study
+         * @return                  interpreted variants.
+         */
+        abstract List<VariantContext> toVariants(final ReferenceMultiSource reference,
+                                                 final Map<String, List<SVFastqUtils.FastqRead>> assemblyID2Reads,
+                                                 final String pathToFastqsDir);
+    }
+
+    /**
+     * Depending on overlapping signatures of input {@code overlappingPair},
+     * make appropriate handler using scenarios recognized in {@link OverlappingScenario}.
+     */
+    private static OverlappingPairHandler factory(final OverlappingPair overlappingPair) {
+
+        final SimpleInterval intervalFive = overlappingPair.fivePrimeBreakEnd.getSpanningInterval();
+        final SimpleInterval intervalThree = overlappingPair.threePrimeBreakEnd.getSpanningInterval();
+
+        if (intervalFive.contains(intervalThree)) {
+            return new FiveContainsThree(overlappingPair);
+        } else if (intervalThree.contains(intervalFive)) {
+            return new ThreeContainsFive(overlappingPair);
+        } else if (compareIntervals(intervalFive, intervalThree) <= 0) {
+            return new FiveInterleavesThree(overlappingPair);
+        } else {
+            return new ThreeInterleavesFive(overlappingPair);
+        }
+    }
+
+    //==================================================================================================================
+
+    /**
+     * Scenario where re-alignment of short reads back to artificial reference haplotypes
+     * is NOT necessary
+     * to resolve which variants are here.
+     */
+    abstract static class NonDegenerateScenario extends OverlappingPairHandler {
```

Comment:
In general I think naming abstract subclasses after their superclass is a good idea for readability, ie call this `NonDegenerateOverlappingPairHandler`